File: program/program/move-all-zeroes-to-end-of-array/MoveAllZeroesToEndOfArray.c
#include <stdio.h>

void moveZeroesToEnd(int arr[], int n) {
    int nonZeroIndex = 0;

    for (int i = 0; i < n; i++) {
        if (arr[i] != 0) {
            arr[nonZeroIndex++] = arr[i];
        }
    }

    
    while (nonZeroIndex < n) {
        arr[nonZeroIndex++] = 0;
    }
}

int main() {
    int arr[] = {1, 0, 4, 2, 0, 5, 0, 3, 6};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Original Array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }

    moveZeroesToEnd(arr, n);

    printf("\nArray after moving zeroes to the end: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }

    return 0;
}

File: program/program/move-all-zeroes-to-end-of-array/moveAllZeroesToEndOfArray.scala
object MoveZeroesToEnd {
  def main(args: Array[String]): Unit = {
    val arr = Array(0, 1, 0, 3, 12)
    println(s"Original array: ${arr.mkString(", ")}")
    
    moveZeroes(arr)
    
    println(s"Array after moving zeroes: ${arr.mkString(", ")}")
  }

  def moveZeroes(arr: Array[Int]): Unit = {
    var nonZeroIndex = 0

    // First, fill in the non-zero elements
    for (i <- arr.indices) {
      if (arr(i) != 0) {
        arr(nonZeroIndex) = arr(i)
        nonZeroIndex += 1
      }
    }

    // Then, fill in the remaining positions with zeroes
    for (i <- nonZeroIndex until arr.length) {
      arr(i) = 0
    }
  }
}
